### PR TITLE
Improve how debug() handles cyclic references

### PR DESCRIPTION
**Pull request by markstory**
Right now if you dump an object that contains cyclic references somewhere in the requested depth the same objects are output multiple times. While this is annoying with small objects it makes reading debug() output for Table objects, or other complex objects really hard as the same content is output multiple times.

Now instead of duplicate content each object is assigned a context 'id'. When an object is encountered after the first time it is output only the short representation is emitted with the reference id. In a separate block of work I want to enhance the HTML output to include links to the original value from the cyclic reference. However, I felt this was a small/useful change on its own.

### CODE REVIEW

File: src/Error/Debugger.php [hunk 1]
```diff
@@ -495,18 +495,19 @@ protected static function _highlight(string $str): string
      */
     public static function exportVar($var, int $depth = 3): string
```
Comment (othercorey) [inline src/Error/Debugger.php]: Can you rename this to `$maxDepth` to represent the value better?

File: src/Error/Debugger.php [hunk 2]
```diff
@@ -522,15 +523,15 @@ protected static function _export($var, int $depth, int $indent): string
 
                 return "'" . $var . "'";
             case 'array':
-                return static::_array($var, $depth - 1, $indent + 1);
+                return static::_array($var, $context->withAddedDepthAndIndent());
```
Comment (othercorey) [inline src/Error/Debugger.php]: This function name `withAddedDepthAndIndent()` feels like you're increasing max depth, not jumping the context to the next depth level. Not sure there's a much better name for it though.

File: tests/TestCase/Error/DebuggerTest.php [hunk 3]
```diff
@@ -310,14 +311,14 @@ public function testExportVar()
 
         $result = Debugger::exportVar($View);
         $expected = <<<TEXT
-object(Cake\View\View) {
-	Html => object(Cake\View\Helper\HtmlHelper) {}
-	Form => object(Cake\View\Helper\FormHelper) {}
+object(Cake\View\View) #0 {
```
Comment (othercorey) [inline tests/TestCase/Error/DebuggerTest.php]: The `#` makes it feel like a count/depth marker not an id marker. Not sure if an `ID` prefix would look better or not.

File: src/Error/Debugger.php (hunk 2 above)
Comment (othercorey) [reply src/Error/Debugger.php]: Maybe `atNextDepthAndIndent()` or `withNextDepthAndIndent()`. Just a thought.

Comment (markstory) [reply src/Error/Debugger.php]: I couldn't come up with a better name. I chose to use the `with` style naming as the DumpContext is a mostly immutable object. I first implemented it as mutable one and tracking depth/indent got really messy.

File: tests/TestCase/Error/DebuggerTest.php (hunk 3 above)
Comment (markstory) [reply tests/TestCase/Error/DebuggerTest.php]: I can do `id:0` if you think it would be clearer.

Comment (othercorey) [reply tests/TestCase/Error/DebuggerTest.php]: Yea, that is what I was thinking.

**Pull request merged**